corehq/apps/registry/models.py

```diff
@@ -124,3 +142,167 @@ class RegistryPermission(models.Model):
+        (ACTION_GRANT_REMOVED, _("Grant removed")),
+        (ACTION_SCHEMA_CHANGED, _("Schema Changed")),
+        (ACTION_DATA_ACCESSED, _("Data Accessed")),
+    )
+
+    RELATED_OBJECT_REGISTRY = "registry"
+    RELATED_OBJECT_INVITATION = "invitation"
+    RELATED_OBJECT_GRANT = "grant"
+    RELATED_OBJECT_UCR = "ucr"
+    RELATED_OBJECT_APPLICATION = "application"  # case search
+    RELATED_OBJECT_CHOICES = (
+        (RELATED_OBJECT_REGISTRY, _("Data Registry")),
+        (RELATED_OBJECT_INVITATION, _("Invitation")),
+        (RELATED_OBJECT_GRANT, _("Grant")),
+        (RELATED_OBJECT_UCR, _("Report")),
+        (RELATED_OBJECT_APPLICATION, _("Case Search")),
+    )
+
+    registry = models.ForeignKey("DataRegistry", related_name="audit_logs", on_delete=models.CASCADE)
+    date = models.DateTimeField(auto_now_add=True)
+    action = models.CharField(max_length=32, choices=ACTION_CHOICES)
+    domain = models.CharField(max_length=255)
+    user = models.ForeignKey(User, related_name="registry_actions", on_delete=models.CASCADE)
+    related_object_id = models.CharField(max_length=32)
+    related_object_type = models.CharField(max_length=32, choices=RELATED_OBJECT_CHOICES)
+    detail = JSONField(null=True)
+
+
+class RegistryAuditHelper:
+    def __init__(self, registry):
+        self.registry = registry
+
+    def registry_activated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=True)
+
+    def registry_deactivated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=False)
+
+    def invitation_accepted(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=True)
+
+    def invitation_rejected(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=False)
+
+    def invitation_added(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=True)
+
+    def invitation_removed(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=False)
+
+    def grant_added(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=True)
+
+    def grant_removed(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=False)
+
+    def schema_changed(self, user, new, old):
+        return RegistryAuditLog.objects.create(
+            registry=self.registry,
+            user=user,
+            action=RegistryAuditLog.ACTION_SCHEMA_CHANGED,
+            domain=self.registry.domain,
+            related_object_id=self.registry.id,
+            related_object_type=RegistryAuditLog.RELATED_OBJECT_REGISTRY,
+            detail={
+                "new_schema": new,
+                "old_schema": old,
+            }
+        )
+
+    def data_accessed(self, user, domain, related_object, filters=None):
```

Comment:
This seems like an extremely high-volume method - will it be used every time any user accesses the registry?  For projects using registries, I was imagining that type of workflow being nearly the same order of magnitude as form submissions.  If so it might be worth considering moving this log model to a sharded DB, or separating this access type from "administrative" type actions and deleting after 90 days or something.

Reply:
This should only be when a user interacts with the registry, so it won't be used when populating UCRs but it will be used when a user views the UCR or performs a case search. The volume will be much higher than the rest but I don't think it will be overwhelming. 

You do bring up a good point about long term maintenance. I wonder if I should partition this by the action type - then we could put these in a separate table.

Reply:
So we're still on PG 9.6 so we can't really take advantage of partitioning properly and the tooling we have only supports range based partitioning. I'm going to punt this for now.